### PR TITLE
feat(binding): Expose subscribing to custom to-device

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Motivation: I am working on a rust rtc stack https://github.com/BillCarsonFr/matrix-rust-rtc. , and some of the needed primitive are missing in rust-sdk.

Some of it was used by the widget driver, but I need it to be exposed to bindings to build the rust-rtc bridge.

- [x] I've documented the public API changes in the appropriate changelog files (see [Writing changelog entries][pull-request-template-changelog]).
- [x] This PR was made with the help of AI.

[pull-request-template-changelog]: https://github.com/matrix-org/matrix-rust-sdk/blob/main/CONTRIBUTING.md#writing-changelog-entries

Signed-off-by:

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 1]
```diff
@@ -2382,6 +2384,87 @@ impl Client {
...
+    type Error = serde_json::Error;
+
+    fn try_from(message: SdkToDeviceMessage) -> Result<Self, Self::Error> {
+        /// The subset of a to-device event that is exposed over FFI.
+        #[derive(Deserialize)]
+        struct ToDeviceMessageHelper<'a> {
+            #[serde(rename = "type")]
+            event_type: String,
+            sender: String,
+            #[serde(borrow)]
```
Comment (andybalaam) [inline bindings/matrix-sdk-ffi/src/client.rs:2460]: Why do we borrow here and then make a copy below?

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 2]
```diff
@@ -2382,6 +2384,87 @@ impl Client {
...
+    /// The listener is called with every to-device message whose type is one
+    /// of `event_types`, or with every custom to-device message if
+    /// `event_types` is empty. A message that was sent encrypted is delivered
+    /// decrypted, along with its encryption info.
+    ///
+    /// The to-device traffic the SDK uses for its own crypto machinery and the
+    /// messages it could not decrypt are never delivered.
+    ///
+    /// Use the returned [`TaskHandle`] to cancel the subscription.
+    pub fn subscribe_to_to_device_messages(
```
Comment (andybalaam) [inline bindings/matrix-sdk-ffi/src/client.rs]: Should this be `subscribe_to_custom_to_device_messages`? (add `custom`). The current name might be confusing for someone trying to see the built-in to-device messages.

Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/client.rs]: (And similar for the similar names below)

File: crates/matrix-sdk/src/client/mod.rs [hunk 3]
```diff
@@ -1410,6 +1416,80 @@ impl Client {
...
+    ///     };
+    ///     println!("{} sent {}", encryption_info.sender, message.raw.json());
+    /// }
+    /// # };
+    /// ```
+    pub fn subscribe_to_to_device_messages(
+        &self,
+        event_types: Vec<ToDeviceEventType>,
+    ) -> impl Stream<Item = ToDeviceMessage> + use<> {
+        let (sender, mut receiver) = unbounded_channel();
```
Comment (andybalaam) [inline crates/matrix-sdk/src/client/mod.rs:1458]: Just for my education, why do we need a channel here instead of directly yielding whenever an event happens?

File: crates/matrix-sdk/tests/integration/encryption/to_device.rs [hunk 4]
```diff
@@ -277,3 +281,50 @@ async fn test_encrypt_and_send_to_device_report_failures_encryption_error() {
...
+        .unwrap()
+        .unwrap();
+
+    let event_synced_future =
+        matrix_mock_server.mock_capture_put_to_device_then_sync_back(bob_user_id, &alice).await;
+
+    bob.encryption()
+        .encrypt_and_send_raw_to_device(
+            vec![&bob_alice_device],
+            "my.custom.to.device",
```
Comment (andybalaam) [inline crates/matrix-sdk/tests/integration/encryption/to_device.rs:312]: Can we send another event type here (or maybe 2: one built-in and one custom?) and confirm that they do not come through?

Can we also have another test that check when you pass an empty vec to subscribe_to_to_device_messages we receive all custom ones and no built-in ones?

Comment (andybalaam) [reply crates/matrix-sdk/tests/integration/encryption/to_device.rs:312]: Ah no, it's fine - I see you covered these cases below.

File: crates/matrix-sdk/tests/integration/to_device.rs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+    server
+        .mock_sync()
+        .ok_and_run(&client, |builder| {
+            builder
+                .add_to_device_event(json!({
+                    "sender": "@alice:example.com",
+                    "type": "m.custom.first",
+                    "content": {},
+                }))
```
Comment (andybalaam) [inline crates/matrix-sdk/tests/integration/to_device.rs:93]: Can we send a built-in event here to show it does not appear?

File: crates/matrix-sdk/tests/integration/to_device.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                "type": "m.custom.wanted",
+                "content": {},
+            }));
+        })
+        .await;
+
+    // A new subscription only sees what arrives after it was created.
+    let stream = client.subscribe_to_to_device_messages(vec![custom("m.custom.wanted")]);
+    pin_mut!(stream);
+    assert_pending!(stream);
```
Comment (andybalaam) [inline crates/matrix-sdk/tests/integration/to_device.rs:222]: Can we confirm that the old message was dropped and the new one was received?

**Review by andybalaam — APPROVED** [review]
Looks sensible to me, with some minor comments. Dropping a note to @poljar just for a quick sanity check - maybe there is some other way to achieve this that we already have? (Note: no need for a full review from poljar - just a check that this is a sensible thing to do)

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 1 above)
Comment (BillCarsonFr) [reply bindings/matrix-sdk-ffi/src/client.rs:2460]: Ok, so my understanding: It cannot be just a `RawJsonValue` because it is not `Sized`.
The compiler is pointing that `Deserialized` is not implemented for `RawJsonValue` but it is for `& RawJsonValue`.
For some reason serde needs the `#[serde(borrow)]` for `RawJsonValue` (it infer it for other simple types though like `&str`).

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 2 above)
Comment (BillCarsonFr) [reply bindings/matrix-sdk-ffi/src/client.rs]: Thx, a lot better.

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 1 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/client.rs:2460]: OK, I don't understand but I don't think it's worth thinking about too much, so long as it works :-)

File: crates/matrix-sdk/src/client/mod.rs (hunk 3 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk/src/client/mod.rs:1458]: For this PR my goal was to extract a helper method that was in `matrix.rs` (widget driver), move it down and then use it in the driver and expose it in the bindings.
So I tried not to touch existing behavior, and this use to be an unbounded_channel in the old code https://github.com/matrix-org/matrix-rust-sdk/blob/c8cb57956ed6f0b28731aeff8c8c9293268b5f98/crates/matrix-sdk/src/widget/matrix.rs#L281

Other than that the unbounded_channel looks like a good fit? send on the unbounded channel always succeed and will buffer if receiver is slow. We don't want to drop to-device on pressure. Not sure what yield would do exactly thouth

File: crates/matrix-sdk/tests/integration/encryption/to_device.rs (hunk 4 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/encryption/to_device.rs:312]: yes, it was a bit more annoying, I had to encrypt them first

File: crates/matrix-sdk/src/client/mod.rs (hunk 3 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/client/mod.rs:1458]: Fair enough to just move around existing code.

File: crates/matrix-sdk/tests/integration/to_device.rs (hunk 5 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/to_device.rs:93]: Updated, and added an assert to check that the content is passed correctly.

File: crates/matrix-sdk/tests/integration/to_device.rs (hunk 6 above)
Comment (BillCarsonFr) [reply crates/matrix-sdk/tests/integration/to_device.rs:222]: Done

**Review by andybalaam — APPROVED** [review]
Great, thanks!

**Comment by andybalaam** [comment]
@BillCarsonFr can you rebase and I will git merge, thanks!

**Comment by BillCarsonFr** [comment]
> @BillCarsonFr can you rebase and I will git merge, thanks!

Should be fine now, I messed up import on rebase, but it is fine now

**Comment by andybalaam** [comment]
> > @BillCarsonFr can you rebase and I will git merge, thanks!
> 
> Should be fine now, I messed up import on rebase, but it is fine now

Can you rebase so there is a clean set of commits so I can fast-forward merge, please? (We don't squash in this repo)

**Comment by BillCarsonFr** [comment]
> > > @BillCarsonFr can you rebase and I will git merge, thanks!
> > 
> > 
> > Should be fine now, I messed up import on rebase, but it is fine now
> 
> Can you rebase so there is a clean set of commits so I can fast-forward merge, please? (We don't squash in this repo)

I re-organized the commits, and I noticed that I forgot to updates the readme after I did the renaming from `subscribe_to_to` to `subscribe_to_custom_` so I updated that. That is the only diff with the old "tip" of the branch

**Comment by andybalaam** [comment]
Argh sorry I made conflicts by merging your other one. One more rebase please?

**Comment by BillCarsonFr** [comment]
> Argh sorry I made conflicts by merging your other one. One more rebase please?

Should be ok now 👍

**Review by andybalaam — APPROVED** [review]
Thanks!
